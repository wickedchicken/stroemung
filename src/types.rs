use crate::math::Real;
use ndarray::{Array, Ix2};

pub type Velocity = [Real; 2];

pub type GridArray<T> = Array<T, Ix2>;

pub type GridSize = [usize; 2];

// According to https://docs.rs/ndarray/0.16.1/ndarray/struct.ArrayBase.html,
// "The default memory order of an array is row major order (a.k.a “c” order),
// where each row is contiguous in memory." This means we can use the default
// implementation of PartialOrd and Ord. If we were using column-major ordering,
// we would need to implement an ordering that prioritized the y value over the
// x value.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct BoundaryIndex(pub usize, pub usize);

// It would be nice to unify BoundaryIndex and GridIndex into one type that
// can be sorted and also directly used by ndarray's indexing operations.
pub type GridIndex = (usize, usize);
