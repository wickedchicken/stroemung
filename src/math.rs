use ndarray::ArrayView2;

pub type Real = f64;

/// Calculate du^2/dx (the derivative of u^2 over x)
///
/// This function uses the same basic algebra rearrangement that the
/// NaSt2D code does. This makes the function easier to compare against
/// NaSt2D (and reduces the number of divisions in the calculation).
///
/// # Arguments
///
/// * `u_view` - A 3x3-element ArrayView2 representing
///   u[(i-1) to (i+1), (j-1) to (j+1)]. This function only uses
///   the three values on the row where j is 0 (index 1), but takes a 3x3
///   ArrayView2 to be easier to combine with other functions.
/// * `delx` - "delta x," the physical width of the cell
/// * `gamma` - Greek letter gamma, the upwind discretization parameter
pub fn du2dx(u_view: ArrayView2<Real>, delx: Real, gamma: Real) -> Real {
    let u_i_m1 = u_view[(0, 1)]; // u[(i-1, j)]  "u[i minus 1]" -> u_i_m1
    let u_i = u_view[(1, 1)]; // u[(i, j)]  "u[i]" -> "u_i"
    let u_i_p1 = u_view[(2, 1)]; // u[(i+1, j)]  "u[i plus 1]" -> u_i_p1

    let inner_left1 = (u_i + u_i_p1).powi(2);
    let inner_right1 = (u_i_m1 + u_i).powi(2);

    let left_side = inner_left1 - inner_right1;

    let inner_left2 = (u_i + u_i_p1).abs() * (u_i - u_i_p1);
    let inner_right2 = (u_i_m1 + u_i).abs() * (u_i_m1 - u_i);

    (left_side + (gamma * (inner_left2 - inner_right2))) / (4.0 * delx)
}

/// Calculate duv/dx (the derivative of u*v over x)
///
/// This function uses the same basic algebra rearrangement that the
/// NaSt2D code does. This makes the function easier to compare against
/// NaSt2D (and reduces the number of divisions in the calculation).
///
/// # Arguments
///
/// * `u_view` - A 3x3-element ArrayView2 representing
///   u[(i-1) to (i+1), (j-1) to (j+1)]. This function only uses the four
///   "lower left" values (the four combinarions of i-1, i, j, and j+1), but
///   takes a 3x3 ArrayView2 to be easier to combine with other functions.
/// * `v_view` - A 3x3-element ArrayView2 representing
///   v[(i-1) to (i+1), (j-1) to (j+1)]. This function only uses
///   the three values on the row where j is 0 (index 1), but takes a 3x3
///   ArrayView2 to be easier to combine with other functions.
/// * `delx` - "delta x," the physical width of the cell
/// * `gamma` - Greek letter gamma, the upwind discretization parameter
pub fn duvdx(
    u_view: ArrayView2<Real>,
    v_view: ArrayView2<Real>,
    delx: Real,
    gamma: Real,
) -> Real {
    let u_i_j = u_view[(1, 1)]; // u[(i, j)] -> u_i_j
    let u_i_j_p1 = u_view[(1, 2)]; // u[(i, j+1)]  "u[i][j plus 1]" -> u_i_j_p1
    let u_i_m1_j = u_view[(0, 1)]; // "u[(i-1, j)]" "u[i minus 1][j]" -> u_i_m1_j
    let u_i_m1_j_p1 = u_view[(0, 2)]; // "u[(i-1, j+1)]" "u[i-1][j+1]" -> u_i_m1_j_p1

    let v_i_j = v_view[(1, 1)]; // "v[(i, j)]" -> v_i_j
    let v_i_p1_j = v_view[(2, 1)]; // "v[(i+1, j)]" -> "v[i plus 1][j]" -> v_i_p1_j
    let v_i_m1_j = v_view[(0, 1)]; // "v[(i-1, j)]" -> "v[i-1][j]" -> v_i_m1_j

    let inner_left1 = (u_i_j + u_i_j_p1) * (v_i_j + v_i_p1_j);
    let inner_right1 = (u_i_m1_j + u_i_m1_j_p1) * (v_i_m1_j + v_i_j);

    let left_side = inner_left1 - inner_right1;

    let inner_left2 = (u_i_j + u_i_j_p1).abs() * (v_i_j - v_i_p1_j);
    let inner_right2 = (u_i_m1_j + u_i_m1_j_p1).abs() * (v_i_m1_j - v_i_j);

    (left_side + (gamma * (inner_left2 - inner_right2))) / (4.0 * delx)
}

/// Calculate duv/dy (the derivative of u*v over y)
///
/// This function uses the same basic algebra rearrangement that the
/// NaSt2D code does. This makes the function easier to compare against
/// NaSt2D (and reduces the number of divisions in the calculation).
///
/// # Arguments
///
/// * `u_view` - A 3x3-element ArrayView2 representing
///   u[(i-1) to (i+1), (j-1) to (j+1)]. This function only uses
///   the three values on the column where i is 0 (index 1), but takes a 3x3
///   ArrayView2 to be easier to combine with other functions.
/// * `v_view` - A 3x3-element ArrayView2 representing
///   v[(i-1) to (i+1), (j-1) to (j+1)]. This function only uses
///   "upper right" values (the four combinarions of i, i+1, j-1, and j), but
///   takes a 3x3 ArrayView2 to be easier to combine with other functions.
/// * `dely` - "delta y," the physical height of the cell
/// * `gamma` - Greek letter gamma, the upwind discretization parameter
pub fn duvdy(
    u_view: ArrayView2<Real>,
    v_view: ArrayView2<Real>,
    dely: Real,
    gamma: Real,
) -> Real {
    let u_i_j = u_view[(1, 1)]; // u[(i, j)] -> u_i_j
    let u_i_j_m1 = u_view[(1, 0)]; // u[(i, j-1)] -> "u[i][j minus 1]" -> u_i_j_m1
    let u_i_j_p1 = u_view[(1, 2)]; // u[(i, j+1)] -> "u[i][j plus 1]" -> u_i_j_p1

    let v_i_j = v_view[(1, 1)]; // v[(i, j)] -> v_i_j
    let v_i_j_m1 = v_view[(1, 0)]; // v[(i, j-1)] -> "v[i][j minus 1]" -> v_i_j_m1
    let v_i_p1_j = v_view[(2, 1)]; // v[(i+1, j)] -> "v[i plus 1][j]" -> v_i_p1_j
    let v_i_p1_j_m1 = v_view[(2, 0)]; // v[(i+1, j-1)] -> "v[i plus 1][j minus 1]" -> v_i_p1_j_m1

    let inner_left1 = (v_i_j + v_i_p1_j) * (u_i_j + u_i_j_p1);
    let inner_right1 = (v_i_j_m1 + v_i_p1_j_m1) * (u_i_j_m1 + u_i_j);
    let left_side = inner_left1 - inner_right1;

    let inner_left2 = (v_i_j + v_i_p1_j).abs() * (u_i_j - u_i_j_p1);
    let inner_right2 = (v_i_j_m1 + v_i_p1_j_m1).abs() * (u_i_j_m1 - u_i_j);

    (left_side + (gamma * (inner_left2 - inner_right2))) / (4.0 * dely)
}

/// Calculate dv^2/dy (the derivative of v^2 over y)
///
/// This function uses the same basic algebra rearrangement that the
/// NaSt2D code does. This makes the function easier to compare against
/// NaSt2D (and reduces the number of divisions in the calculation).
///
/// # Arguments
///
/// * `v_view` - A 3x3-element ArrayView2 representing
///   v[(i-1) to (i+1), (j-1) to (j+1)]. This function only uses
///   the three values on the column where i is 0 (index 1), but takes a 3x3
///   ArrayView2 to be easier to combine with other functions.
/// * `dely` - "delta y," the physical width of the cell
/// * `gamma` - Greek letter gamma, the upwind discretization parameter
pub fn dv2dy(v_view: ArrayView2<Real>, dely: Real, gamma: Real) -> Real {
    let v_i_j = v_view[(1, 1)]; // v[(i, j)] -> v_i_j
    let v_i_j_p1 = v_view[(1, 2)]; // v[(i, j+1)] -> "v[i][j plus 1]" -> v_i_j_p1
    let v_i_j_m1 = v_view[(1, 0)]; // v[(i, j-1)] -> "v[i][j minus 1]" -> v_i_j_m1

    let inner_left1 = (v_i_j + v_i_j_p1).powi(2);
    let inner_right1 = (v_i_j_m1 + v_i_j).powi(2);

    let left_side = inner_left1 - inner_right1;

    let inner_left2 = (v_i_j + v_i_j_p1).abs() * (v_i_j - v_i_j_p1);
    let inner_right2 = (v_i_j_m1 + v_i_j).abs() * (v_i_j_m1 - v_i_j);

    (left_side + (gamma * (inner_left2 - inner_right2))) / (4.0 * dely)
}

#[cfg(test)]
mod tests {
    use super::*;
    use ndarray::{array, ArrayView2};

    #[test]
    fn test_du2dx() {
        // These don't have any particular significance, just some random data.
        let test_cases = [
            (
                array![[0., 1., 0.], [0., 2., 0.], [0., 3., 0.]],
                1.,
                1.7,
                3.15,
            ),
            (
                array![[0., -1., 0.], [0., 2., 0.], [0., 3., 0.]],
                1.,
                1.7,
                5.15,
            ),
            (
                array![[0., 1., 0.], [0., 1., 0.], [0., 1., 0.]],
                1.,
                1.7,
                0.,
            ),
            (
                array![[0., 1., 0.], [0., 2., 0.], [0., 3., 0.]],
                1.,
                1.3,
                3.35,
            ),
            (
                array![[0., 1., 0.], [0., 2., 0.], [0., 3., 0.]],
                1.5,
                1.7,
                2.1,
            ),
            (
                array![[0., 10., 0.], [0., 20., 0.], [0., 30., 0.]],
                1.5,
                1.7,
                210.,
            ),
        ];
        for (u, delx, gamma, expected) in test_cases {
            assert_eq!(du2dx(ArrayView2::from(&u), delx, gamma), expected);
        }
    }

    #[test]
    fn test_duvdx() {
        // These don't have any particular significance, just some random data.
        let test_cases = [
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.,
                1.7,
                40.35,
            ),
            (
                array![[1., 2., 3.], [4., 5., -6.], [-7., 8., 9.]],
                array![[8., 9., 10.], [11., -12., 13.], [14., 15., -16.]],
                1.,
                1.7,
                -53.1,
            ),
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.6,
                1.7,
                25.218750,
            ),
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.,
                1.5,
                41.25,
            ),
        ];
        for (u, v, delx, gamma, expected) in test_cases {
            assert_eq!(
                duvdx(ArrayView2::from(&u), ArrayView2::from(&v), delx, gamma),
                expected
            );
        }
    }

    #[test]
    fn test_duvdy() {
        // These don't have any particular significance, just some random data.
        let test_cases = [
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.,
                1.7,
                17.15,
            ),
            (
                array![[1., 2., 3.], [4., 5., -6.], [-7., 8., 9.]],
                array![[8., 9., 10.], [11., -12., 13.], [14., 15., -16.]],
                1.,
                1.7,
                -32.35,
            ),
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.6,
                1.7,
                10.718749999999998,
            ),
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.,
                1.5,
                17.25,
            ),
        ];
        for (u, v, dely, gamma, expected) in test_cases {
            assert_eq!(
                duvdy(ArrayView2::from(&u), ArrayView2::from(&v), dely, gamma),
                expected
            );
        }
    }

    #[test]
    fn test_dv2dy() {
        // These don't have any particular significance, just some random data.
        let test_cases = [
            (
                array![[0., 0., 0.], [1., 2., 3.], [0., 0., 0.]],
                1.,
                1.7,
                3.15,
            ),
            (
                array![[0., 0., 0.], [-1., 2., 3.], [0., 0., 0.]],
                1.,
                1.7,
                5.15,
            ),
            (
                array![[0., 0., 0.], [1., 1., 1.], [0., 0., 0.]],
                1.,
                1.7,
                0.,
            ),
            (
                array![[0., 0., 0.], [1., 2., 3.], [0., 0., 0.]],
                1.,
                1.3,
                3.35,
            ),
            (
                array![[0., 0., 0.], [1., 2., 3.], [0., 0., 0.]],
                1.5,
                1.7,
                2.1,
            ),
            (
                array![[0., 0., 0.], [10., 20., 30.], [0., 0., 0.]],
                1.5,
                1.7,
                210.,
            ),
        ];
        for (v, dely, gamma, expected) in test_cases {
            assert_eq!(dv2dy(ArrayView2::from(&v), dely, gamma), expected);
        }
    }
}
