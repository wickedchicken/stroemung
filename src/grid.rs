use std::fmt;

use ndarray::Array;
use serde::Deserialize;
use serde::Serialize;

use crate::math::Real;
use crate::types::{GridArray, GridSize};

#[derive(Debug, Serialize, Deserialize)]
pub struct SimulationGrid {
    size: GridSize,
    pressure: GridArray<Real>,
}

impl SimulationGrid {
    pub fn new(size: GridSize) -> SimulationGrid {
        SimulationGrid {
            size,
            pressure: Array::zeros(size),
        }
    }
}

impl std::fmt::Display for SimulationGrid {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "Simulation grid {}x{}", self.size[0], self.size[1],)?;
        writeln!(f, "Pressure:{}", self.pressure)?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn grid_size() {
        let size = [5, 7];
        let grid = SimulationGrid::new(size);
        assert_eq!(grid.size[0], 5);
        assert_eq!(grid.size[1], 7);
        assert_eq!(grid.pressure.shape(), size);
    }

    #[test]
    fn serialize() {
        let size = [2, 3];
        let grid = SimulationGrid::new(size);
        let serialized = serde_json::to_string(&grid).unwrap();
        assert_eq!(
            serialized,
            "{\"size\":[2,3],\"pressure\":{\"v\":1,\"dim\":[2,3],\"data\":[0.0,0.0,0.0,0.0,0.0,0.0]}}"
        );
    }
}
