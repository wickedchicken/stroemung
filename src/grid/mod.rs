use std::fmt;

use ndarray::Array;
use serde::Deserialize;
use serde::Serialize;

use crate::math::Real;
use crate::types::{GridArray, GridSize};

#[derive(Debug, Serialize, Deserialize)]
pub struct SimulationGrid {
    size: GridSize,
    pressure: GridArray<Real>,
}

impl SimulationGrid {
    pub fn new(size: GridSize) -> SimulationGrid {
        SimulationGrid {
            size,
            pressure: Array::zeros(size),
        }
    }
}

impl std::fmt::Display for SimulationGrid {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "Simulation grid {}x{}", self.size[0], self.size[1],)?;
        writeln!(f, "Pressure:{}", self.pressure)?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn grid_size() {
        let size = [5, 7];
        let grid = SimulationGrid::new(size);
        assert_eq!(grid.size[0], 5);
        assert_eq!(grid.size[1], 7);
        assert_eq!(grid.pressure.shape(), size);
    }

    #[test]
    fn deserialize() {
        use std::fs::File;
        use std::io::BufReader;
        use std::path::Path;

        let test_data_directory = Path::new(file!())
            .parent()
            .unwrap()
            .parent()
            .unwrap()
            .join("test_data");
        let test_filename = test_data_directory.join("simple_grid.json");
        let result: SimulationGrid =
            serde_json::from_reader(BufReader::new(File::open(test_filename).unwrap()))
                .unwrap();
        insta::assert_json_snapshot!(result);
    }

    #[test]
    fn serialize() {
        let size = [2, 3];
        let grid = SimulationGrid::new(size);
        insta::assert_json_snapshot!(grid);
    }
}
