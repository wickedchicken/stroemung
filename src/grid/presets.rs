use crate::cell::Cell;
use crate::grid::{SimulationGrid, UnfinalizedSimulationGrid};
use crate::types::GridSize;
use ndarray::Array;

/// Generate an empty simulation grid
pub fn empty(size: GridSize) -> SimulationGrid {
    SimulationGrid::try_from(UnfinalizedSimulationGrid {
        size,
        pressure: Array::zeros(size),
        u: Array::zeros(size),
        v: Array::zeros(size),
        cell_type: Array::from_elem(size, Cell::Fluid),
    })
    .unwrap()
}
