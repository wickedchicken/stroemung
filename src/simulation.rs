use std::error::Error;
use std::fmt;
use std::io::Read;

use crate::math::Real;
use crate::math::{du2dx, duvdx, duvdy, dv2dy, laplacian};

use serde::Deserialize;
use serde::Serialize;

use crate::grid::{SimulationGrid, UnfinalizedSimulationGrid};
use crate::types::{CellPhysicalSize, GridSize};

use ndarray::ArrayView2;

#[derive(Debug, Deserialize)]
pub struct UnfinalizedSimulation {
    size: GridSize,
    cell_size: CellPhysicalSize,
    delt: Real,
    gamma: Real,
    reynolds: Real,
    grid: UnfinalizedSimulationGrid,
}

// This must be the same as UnfinalizedSimulation, except the type
// of grid and without the calculated values. We have two types to make sure
// we never deserialize without forgetting to generate the boundary list.
#[derive(Debug, Serialize)]
pub struct Simulation {
    size: GridSize,
    cell_size: CellPhysicalSize,
    delt: Real,
    gamma: Real,
    reynolds: Real,
    grid: SimulationGrid,
}

impl From<UnfinalizedSimulation> for Simulation {
    fn from(item: UnfinalizedSimulation) -> Self {
        // Will be nicer once https://github.com/rust-lang/rust/issues/86555
        // is in stable.
        Simulation {
            size: item.size,
            cell_size: item.cell_size,
            delt: item.delt,
            gamma: item.gamma,
            reynolds: item.reynolds,
            grid: item.grid.into(),
        }
    }
}

impl std::fmt::Display for Simulation {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(
            f,
            "Simulation {}x{} cells, {}x{} physical size",
            self.size[0], self.size[1], self.cell_size[0], self.cell_size[1]
        )?;
        writeln!(f, "Time step delta:{}", self.delt)?;
        writeln!(f, "Gamma:{}", self.gamma)?;
        writeln!(f, "Reynolds number:{}", self.reynolds)?;
        writeln!(f, "{}", self.grid)?;
        Ok(())
    }
}

impl Simulation {
    pub fn from_reader<R: Read>(reader: R) -> Result<Simulation, Box<dyn Error>> {
        let unfinalized: UnfinalizedSimulation = serde_json::from_reader(reader)?;
        Ok(Simulation::from(unfinalized))
    }
}

/// Calculate F (the horizontal non-pressure part of the momentum equation)
///
/// # Arguments
///
/// * `u_view` - A 3x3-element ArrayView2 representing
///   u[(i-1) to (i+1), (j-1) to (j+1)].
/// * `v_view` - A 3x3-element ArrayView2 representing
///   v[(i-1) to (i+1), (j-1) to (j+1)].
/// * `delx` - "delta x," the physical width of the cell
/// * `dely` - "delta y," the physical width of the cell
/// * `delt` - "delta t," the amount of time per time step
/// * `gamma` - Greek letter gamma, the upwind discretization parameter
/// * `reynolds` - The Reynolds number for the simulation
pub fn calculate_f(
    u_view: ArrayView2<Real>,
    v_view: ArrayView2<Real>,
    delx: Real,
    dely: Real,
    delt: Real,
    gamma: Real,
    reynolds: Real,
) -> Real {
    u_view[(1, 1)]
        + (delt
            * ((laplacian(u_view, delx, dely) / reynolds)
                - du2dx(u_view, delx, gamma)
                - duvdy(u_view, v_view, dely, gamma)))
}

/// Calculate G (the vertical non-pressure part of the momentum equation)
///
/// # Arguments
///
/// * `u_view` - A 3x3-element ArrayView2 representing
///   u[(i-1) to (i+1), (j-1) to (j+1)].
/// * `v_view` - A 3x3-element ArrayView2 representing
///   v[(i-1) to (i+1), (j-1) to (j+1)].
/// * `delx` - "delta x," the physical width of the cell
/// * `dely` - "delta y," the physical width of the cell
/// * `delt` - "delta t," the amount of time per time step
/// * `gamma` - Greek letter gamma, the upwind discretization parameter
/// * `reynolds` - The Reynolds number for the simulation
pub fn calculate_g(
    u_view: ArrayView2<Real>,
    v_view: ArrayView2<Real>,
    delx: Real,
    dely: Real,
    delt: Real,
    gamma: Real,
    reynolds: Real,
) -> Real {
    v_view[(1, 1)]
        + (delt
            * ((laplacian(v_view, delx, dely) / reynolds)
                - duvdx(u_view, v_view, delx, gamma)
                - dv2dy(v_view, dely, gamma)))
}

#[cfg(test)]
mod tests {
    use super::*;
    use ndarray::{array, ArrayView2};
    use std::fs::File;
    use std::io::BufReader;
    use std::path::{Path, PathBuf};

    use crate::grid::presets;

    fn test_data_directory() -> PathBuf {
        Path::new(file!()).parent().unwrap().join("test_data")
    }

    #[test]
    fn deserialize() {
        let test_filename = test_data_directory().join("simple_simulation.json");
        let result =
            Simulation::from_reader(BufReader::new(File::open(test_filename).unwrap()))
                .unwrap();
        insta::assert_json_snapshot!(result);
    }

    #[test]
    fn serialize() {
        let size = [5, 7];
        let cell_size = [1., 2.];
        let delt = 1.4;
        let gamma = 1.7;
        let reynolds = 100.;

        let simulation = Simulation::from(UnfinalizedSimulation {
            size,
            cell_size,
            delt,
            gamma,
            reynolds,
            grid: presets::empty(size).into(),
        });

        insta::assert_json_snapshot!(simulation);
    }

    #[test]
    fn test_calculate_f() {
        // These don't have any particular significance, just some random data.
        let test_cases = [
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.,
                1.,
                0.005,
                1.7,
                100.,
                4.802500,
            ),
            (
                array![[1., 2., 3.], [4., 5., -6.], [-7., 8., 9.]],
                array![[8., 9., 10.], [11., -12., 13.], [14., 15., -16.]],
                1.,
                1.,
                0.006,
                1.7,
                10.,
                5.052800,
            ),
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.6,
                1.,
                0.007,
                1.7,
                14.,
                4.782168750,
            ),
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.,
                1.6,
                0.003,
                1.5,
                400.,
                4.89790625,
            ),
        ];
        for (u, v, delx, dely, delt, gamma, reynolds, expected) in test_cases {
            assert_eq!(
                calculate_f(
                    ArrayView2::from(&u),
                    ArrayView2::from(&v),
                    delx,
                    dely,
                    delt,
                    gamma,
                    reynolds,
                ),
                expected
            );
        }
    }

    #[test]
    fn test_calculate_g() {
        // These don't have any particular significance, just some random data.
        let test_cases = [
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.,
                1.,
                0.005,
                1.7,
                100.,
                11.6825,
            ),
            (
                array![[1., 2., 3.], [4., 5., -6.], [-7., 8., 9.]],
                array![[8., 9., 10.], [11., -12., 13.], [14., 15., -16.]],
                1.,
                1.,
                0.006,
                1.7,
                10.,
                -11.5014,
            ),
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.6,
                1.,
                0.007,
                1.7,
                14.,
                11.66141875,
            ),
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.,
                1.6,
                0.003,
                1.5,
                400.,
                11.83265625,
            ),
        ];
        for (u, v, delx, dely, delt, gamma, reynolds, expected) in test_cases {
            assert_eq!(
                calculate_g(
                    ArrayView2::from(&u),
                    ArrayView2::from(&v),
                    delx,
                    dely,
                    delt,
                    gamma,
                    reynolds,
                ),
                expected
            );
        }
    }
}
