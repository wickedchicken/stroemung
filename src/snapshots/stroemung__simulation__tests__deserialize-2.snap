---
source: src/simulation.rs
expression: result
---
{
  "size": [
    4,
    3
  ],
  "cell_size": [
    0.1,
    0.2
  ],
  "delt": 0.005,
  "gamma": 0.9,
  "reynolds": 100.0,
  "initial_norm_squared": 899.9547140394143,
  "sor_absolute_epsilon": 0.001,
  "max_iterations": 100,
  "iterations": 0,
  "time": 0.0,
  "omega": 1.7,
  "grid": {
    "size": [
      4,
      3
    ],
    "pressure": {
      "v": 1,
      "dim": [
        4,
        3
      ],
      "data": [
        0.0,
        0.2539642899437815,
        0.0,
        0.2539642899437815,
        0.2539669433626628,
        0.2539642899437815,
        -0.1460294667176743,
        -0.146030992433531,
        -0.1460294667176743,
        0.0,
        -0.1460294667176743,
        0.0
      ]
    },
    "u": {
      "v": 1,
      "dim": [
        4,
        3
      ],
      "data": [
        -0.999998967898097,
        1.0,
        -0.999998967898097,
        -0.999998967898097,
        0.999998967898097,
        -0.999998967898097,
        -0.999998967898097,
        0.999998967898097,
        -0.999998967898097,
        -1.0,
        1.0,
        -1.0
      ]
    },
    "v": {
      "v": 1,
      "dim": [
        4,
        3
      ],
      "data": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ]
    },
    "cell_type": {
      "v": 1,
      "dim": [
        4,
        3
      ],
      "data": [
        {
          "Boundary": "NoSlip"
        },
        {
          "Boundary": {
            "Inflow": {
              "velocity": [
                1.0,
                0.0
              ]
            }
          }
        },
        {
          "Boundary": "NoSlip"
        },
        {
          "Boundary": "NoSlip"
        },
        "Fluid",
        {
          "Boundary": "NoSlip"
        },
        {
          "Boundary": "NoSlip"
        },
        "Fluid",
        {
          "Boundary": "NoSlip"
        },
        {
          "Boundary": "NoSlip"
        },
        {
          "Boundary": "Outflow"
        },
        {
          "Boundary": "NoSlip"
        }
      ]
    }
  }
}
